decompiler/structures/visitors/substitute_visitor.py

```diff
@@ -0,0 +1,152 @@
+        return self._visit_operation(op)
+
+    def visit_unary_operation(self, op: UnaryOperation) -> Optional[DataflowObject]:
+        if op.array_info is not None:
+            if (base_replacement := op.array_info.base.accept(self)) is not None:
+                op.array_info.base = _assert_type(base_replacement, Variable)
+            if (isinstance(op.array_info.index, Variable) and
+                    (index_replacement := op.array_info.index.accept(self)) is not None):
+                op.array_info.index = _assert_type(index_replacement, Variable)
+
+        return self._visit_operation(op)
+
+    def visit_binary_operation(self, op: BinaryOperation) -> Optional[DataflowObject]:
+        return self._visit_operation(op)
+
+    def visit_call(self, op: Call) -> Optional[DataflowObject]:
+        if (function_replacement := op.function.accept(self)) is not None:
+            op._function = _assert_type(
+                function_replacement,
+                Union[FunctionSymbol, ImportedFunctionSymbol, IntrinsicSymbol, Variable]
+            )
+
+        return self._visit_operation(op)
+
+    def visit_condition(self, op: Condition) -> Optional[DataflowObject]:
+        return self._visit_operation(op)
+
+    def visit_ternary_expression(self, op: TernaryExpression) -> Optional[DataflowObject]:
+        return self._visit_operation(op)
+
+    def visit_comment(self, instr: Comment) -> Optional[DataflowObject]:
+        return self._mapper(instr)
+
+    def visit_assignment(self, instr: Assignment) -> Optional[DataflowObject]:
+        if (value_replacement := instr.value.accept(self)) is not None:
+            instr._value = _assert_type(value_replacement, Expression)
+        if (destination_replacement := instr.destination.accept(self)) is not None:
+            instr._destination = _assert_type(destination_replacement, Expression)
+
+        return self._mapper(instr)
+
+    def visit_generic_branch(self, instr: GenericBranch) -> Optional[DataflowObject]:
+        if (condition_replacement := instr.condition.accept(self)) is not None:
+            instr._condition = _assert_type(condition_replacement, Expression)
+
+        return self._mapper(instr)
+
+    def visit_return(self, instr: Return) -> Optional[DataflowObject]:
+        if (values_replacement := instr.values.accept(self)) is not None:
+            instr._values = _assert_type(values_replacement, ListOperation)
+
+        return self._mapper(instr)
+
+    def visit_break(self, instr: Break) -> Optional[DataflowObject]:
+        return self._mapper(instr)
+
+    def visit_continue(self, instr: Continue) -> Optional[DataflowObject]:
+        return self._mapper(instr)
+
+    def visit_phi(self, instr: Phi) -> Optional[DataflowObject]:
+        # we ignore the return value here, because replacing instr.value itself would require updating
+        # instr.origin_block with information we don't have
+        instr.value.accept(self)
```

Comment:
The reason this code seems a little bit weird is because the original substitute method on `DataflowObject` and my code tried to disallow any changes which would create 'invalid states'. Like for example when a phi function references a variable from another basic block, but `phi.origin_block` doesn't reference this variable.
These checks are very brittle and impossible to do 100% correctly from within the substitute visitor. So I changed the `SubstistuteVisitor` to just simply substitutes every encountered object in the dataflow tree strict by its `._mapper` and pass the responsibility of not creating 'invalid states' to the user of the visitor.
See f26128b7.